toolkit/tools/internal/retry/retry_test.go

```diff
@@ -0,0 +1,83 @@
+package retry
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// Delay vs. retry count for mult = 1.0, delay = .1 seconds
+var expectedValuesMillis1 = []int{0, 100, 200, 300, 400, 500}
+
+// Delay vs. retry count for exp = 2.0, delay = 0.1 seconds
+var expectedValuesMillis2 = []int{0, 100, 400, 900, 1600, 2500}
+
+func TestCalcDelay1(t *testing.T) {
+	mult := 1.0
+	sleep := time.Millisecond * 100
+	for i, expected := range expectedValuesMillis1 {
+		expected := time.Duration(expected) * time.Millisecond
+		actual := calcDelay(i, sleep, mult)
+		assert.Equal(t, expected, actual)
+	}
+}
+
+func TestCalcDelay2(t *testing.T) {
+	mult := 2.0
+	sleep := time.Millisecond * 100
+	for i, expected := range expectedValuesMillis2 {
+		expected := time.Duration(expected) * time.Millisecond
+		actual := calcDelay(i, sleep, mult)
+		assert.Equal(t, expected, actual)
+	}
+}
+
+func TestTotalRunTimeWithFailures1(t *testing.T) {
+	startTime := time.Now()
+	err := RunWithExpBackoff(func() error {
+		return fmt.Errorf("test error")
+	}, 3, time.Millisecond*100, 1.0)
+	endTime := time.Now()
+	assert.NotNil(t, err)
+	// Delays should be 0 seconds, <attempt>, .1 second, <attempt>, .2 seconds, <attempt>
+	// for a total of .3 seconds.
+	minDelay := time.Millisecond * 300
+	maxDelay := time.Millisecond * 500
```

Comment:
It isn't immediately clear that this represents a 200ms buffer for the timer expiry.

Perhaps, something like:

```golang
expectedDelay := time.Millisecond * 300
delayBuffer := time.Millisecond * 200

assert.GreaterOrEqual(t, endTime.Sub(startTime), expectedDelay)
assert.LessOrEqual(t, endTime.Sub(startTime), expectedDelay + delayBuffer)
```